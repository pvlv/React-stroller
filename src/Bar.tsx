import * as React from 'react';
import {axisTypes} from "./utils";

export type BarView = React.ComponentType<{ dragging?: boolean }> | React.ComponentType<any>;

export type BarSizeFunction = (height: number, scrollHeight: number, flags: { dragging: boolean }) => number;

export interface StrollerBarProps {
  scrollSpace: number;
  space: number;
  forwardRef: (ref: HTMLElement) => void;
  internal?: BarView;
  axis?: axisTypes,
  oppositePosition?: boolean;
  draggable?: boolean;
  dragging?: boolean;
  sizeFunction?: BarSizeFunction,
  barTransform: string
}

const Bar: BarView = () => (
  <div
    style={{
      width: '8px',
      height: '100%',
      borderRadius: 8,
      backgroundColor: 'rgba(0,0,0,0.5)'
    }}
  />
);

const positions = {
  vertical: {
    0: {
      top: 0,
      right: 0,
    },
    1: {
      top: 0,
      left: 0,
    }
  },

  horizontal: {
    0: {
      bottom: 0,
      left: 0,
    },
    1: {
      top: 0,
      left: 0,
    },
  }
};

export const defaultSizeFunction = (height: number, scrollHeight: number): number => (
  height * (height / scrollHeight)
);

export const StollerBar: React.SFC<StrollerBarProps> = ({
                                                          scrollSpace,
                                                          space,
                                                          forwardRef,
                                                          internal,
                                                          axis = 'vertical',
                                                          oppositePosition = false,
                                                          draggable = false,
                                                          sizeFunction = defaultSizeFunction,
                                                          dragging = false,
                                                          barTransform
                                                        }) => {
  if (scrollSpace <= space) {
    return null;
  }

  const barHeight = sizeFunction(space, scrollSpace, {dragging});

  const Internal: BarView = internal || Bar;

  return (
    <div
      ref={forwardRef as any}
      style={{
        position: 'absolute',
        display: 'flex',
        cursor: dragging ? 'grabbing' : (draggable ? 'grab' : 'default'),

        [axis == 'vertical' ? 'height' : 'width']: Math.round(barHeight) + 'px',

        ...(positions[axis][oppositePosition ? 1 : 0] as any),

        transform: barTransform + ' translateZ(0)',
        willChange: 'transform'
      }}
    >
      <Internal dragging={dragging}/>
    </div>
  );
}