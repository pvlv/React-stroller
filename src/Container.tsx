import * as React from 'react';
import {axisToOverflow, axisTypes, getScrollBarWidth} from "./utils";

export interface IContainerProps {
  axis?: axisTypes;
  className?: string;
  overscroll?: boolean;
  gap?: number;
  containerStyles?: CSSStyleDeclaration;
}

const getStyle = (scrollWidth: number, gap: number, overscroll: boolean, axis: axisTypes = 'vertical'): React.CSSProperties => {
  return {
    width: axis === 'vertical' ? `calc(100% + ${scrollWidth - gap}px)` : '100%',
    height: axis !== 'vertical' ? `calc(100% + ${scrollWidth - gap}px)` : '100%',
    // width:'100%',
    // height:'100%',
    maxWidth: 'inherit',
    maxHeight: 'inherit',
    position: 'relative',
    [axisToOverflow[axis]]: 'scroll',
    overscrollBehavior: overscroll ? 'contain' : 'inherit',
    [axis === 'vertical' ? 'paddingRight' : 'paddingBottom']: (scrollWidth + 24) + 'px',
    boxSizing: "content-box",
  }
};

const containerStyle: React.CSSProperties = {
  height: '100%',
  width: '100%',
  overflow: 'hidden',
  maxWidth: 'inherit',
  maxHeight: 'inherit',
};

export const strollerStyle: React.CSSProperties = {
  height: '100%',
  width: '100%',
  maxWidth: 'inherit',
  maxHeight: 'inherit',
  display: 'inline-block',
};

export const subcontainerStyle: React.CSSProperties = {
  // minHeight: '100%', // an issue for windows
  // minWidth: '100%',
  width: '100%',
  height: '100%',
  // maxWidth: 'inherit',
  // maxHeight: 'inherit',
  position: 'relative',
  // display: 'inline-block', // WHY?
};

export class Strollable extends React.Component<IContainerProps> {
  scrollWidth = getScrollBarWidth();

  render() {
    const {children, axis, overscroll = false, className, gap = 0, containerStyles = {}} = this.props;
    return (
      <div style={containerStyle} className={className}>
        <div style={getStyle(this.scrollWidth, gap, overscroll, axis)}>
          <div style={{...subcontainerStyle, ...containerStyles}}>
            {children}
          </div>
        </div>
      </div>
    );
  }
}