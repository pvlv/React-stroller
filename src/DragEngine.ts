import {AnyHookCallback, faste, HookCallback, Faste, InternalMachine, MessageHandler} from 'faste';

export {
  Faste,
  InternalMachine,
  MessageHandler
};

const nodeHook: AnyHookCallback = {
  on: ({attrs, trigger, message}) => {
    const hook = (event: any) => {
      trigger(message, event);
      event.preventDefault();
    };
    attrs.node.addEventListener(message, hook);
    return [attrs.node, hook]
  },
  off: ({message}, [node, hook]) => {
    node.removeEventListener(message, hook)
  }
};

const documentHook: HookCallback<any, any, any> = {
  on: ({trigger, message}) => {
    const hook = (event: any) => {
      trigger(message, event);
      event.preventDefault();
    };
    document.addEventListener(message, hook, true);
    return hook
  },
  off: ({message}, hook) => {
    document.removeEventListener(message, hook, true)
  }
};

const getCoords = (event: MouseEvent | Touch) => [event.clientX, event.clientY];

export const DragMachine = faste()
  .withPhases(['init', 'disabled', 'idle', 'dragging', 'cancelDrag'])
  .withAttrs<{ node?: HTMLElement, enabled?: boolean }>({})
  .withMessages(['check', 'down', 'up', 'move', 'mousedown', 'mouseup', 'mousemove', 'touchstart', 'touchmove', 'touchend'])
  .withSignals(['up', 'down', 'move'])

  .on('check', ['init', 'disabled'], ({attrs, transitTo}) => attrs.enabled && attrs.node && transitTo('idle'))
  .on('check', ['idle', 'dragging'], ({attrs, transitTo}) => (!attrs.enabled || !attrs.node) && transitTo('disabled'))

  // outer reactions
  .on('down', ({transitTo, emit}, event) => {
    emit('down', event);
    transitTo('dragging')
  })
  .on('up', ({transitTo}) => transitTo('idle'))
  .on('move', ({emit}, event) => emit('move', event))
  .on('@enter', ['cancelDrag'], ({transitTo}) => transitTo('idle'))

  // mouse events
  .on('mousedown', ['idle'], ({trigger}, event: MouseEvent) => trigger('down', getCoords(event)))
  .on('mouseup', ['dragging'], ({trigger}) => trigger('up'))
  .on('mousemove', ['dragging'], ({transitTo}, event: MouseEvent) => event.buttons !== 1 && transitTo('cancelDrag'))
  .on('mousemove', ['dragging'], ({trigger}, event: MouseEvent) => trigger('move', getCoords(event)))

  // touch events
  .on('touchstart', ['idle'], ({trigger}, event: TouchEvent) => trigger('down', getCoords(event.touches[0])))
  .on('touchend', ['dragging'], ({trigger}) => trigger('up'))
  .on('touchmove', ['dragging'], ({trigger}, event: TouchEvent) => trigger('move', getCoords(event.touches[0])))

  .hooks({
    mousedown: nodeHook,
    mouseup: documentHook,
    mousemove: documentHook,

    touchstart: nodeHook,
    touchmove: documentHook,
    touchend: documentHook,
  });