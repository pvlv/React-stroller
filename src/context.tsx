import * as React from 'react';

export interface IStrollerContext {
  setScrollContainer: (ref: HTMLElement | null) => any;
}

export const context = React.createContext({
  setScrollContainer: () => { throw new Error('StrollerCaptor used without Stroller')}
} as IStrollerContext);

export const StrollerProvider = context.Provider;
export const StrollerContext = context.Consumer;