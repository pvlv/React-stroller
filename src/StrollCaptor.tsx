import * as React from 'react';
import {StrollerContext} from "./context";

export const StrollCaptor = () => (
  <StrollerContext>
    {({setScrollContainer}) => (
      <div ref={ref => setScrollContainer(ref)}/>
    )}
  </StrollerContext>
);