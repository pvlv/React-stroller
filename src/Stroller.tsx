import * as React from 'react';

import {axisToProps, axisTypes, extractValues, findScrollableParent} from "./utils";
import {BarLocation, BarSizeFunction, BarView, StollerBar, IStrollerBarProps, ISideBar} from "./Bar";
import {DragMachine} from "./DragEngine";

import {StrollerProvider, StrollerStateProvider} from './context';
import {strollerStyle} from "./Container";

export interface IStrollerProps {
  axis?: axisTypes;
  targetAxis?: axisTypes;

  bar?: BarView,
  scrollBar?: React.ComponentType<IStrollerBarProps>;
  barSizeFunction?: BarSizeFunction;
  barClassName?: string,
  SideBar?: ISideBar;

  oppositePosition?: boolean,
  draggable?: boolean

  overrideLocation?: BarLocation;

  scrollKey?: any;
}

export interface IComponentState {
  scrollWidth: number;
  scrollHeight: number;

  clientWidth: number;
  clientHeight: number;

  scrollLeft: number;
  scrollTop: number;

  hasScroll: boolean,

  dragPhase: string;
  mousePosition: number[];
  scrollPosition: number[];

  barLocation: BarLocation;
}

export class Stroller extends React.Component<IStrollerProps, IComponentState> {

  state = {
    scrollWidth: 0,
    scrollHeight: 0,
    clientWidth: 0,
    clientHeight: 0,
    scrollLeft: 0,
    scrollTop: 0,
    dragPhase: 'idle',
    mousePosition: [0, 0],
    scrollPosition: [0, 0],
    barLocation: 'inside' as BarLocation,
    hasScroll: false,
  };

  private dragMachine = DragMachine.create();

  private topNode: HTMLElement | undefined = undefined;
  private scrollableParent: HTMLElement | undefined = undefined;
  private scrollContainer: HTMLElement | null = null;
  private barRef: HTMLElement | undefined = undefined;
  private attachParent: HTMLElement | Window | undefined = undefined;

  componentDidMount() {
    this.scrollableParent = findScrollableParent(this.scrollContainer || this.topNode!, this.props.axis);
    const scrollableParent: any = this.scrollableParent;

    const barLocation = this.props.overrideLocation || this.scrollableParent === document.body
      ? 'fixed'
      : (
        (this.scrollContainer ? !this.topNode!.contains(this.scrollableParent) : true)
          ? 'inside'
          : 'outside'
      );

    this.setState({
      barLocation
    });

    this.attach(barLocation === 'fixed' ? window : this.scrollableParent);

    this.onContainerScroll();

    (this.dragMachine as any)._id = this;

    this.dragMachine
      .attrs({enabled: this.props.draggable})
      .observe((dragPhase: string) => this.setState({dragPhase}))
      .connect((message: string, coords: number[]) => {
        if (message === 'down') {
          this.setState({
            mousePosition: coords,
            scrollPosition:
              this.state.barLocation === 'fixed'
                ? [window.scrollX, window.scrollY]
                : [scrollableParent.scrollLeft, scrollableParent.scrollTop]
          })
        }
        if (message === 'move') {
          const {axis = 'vertical', targetAxis: pTargetAxis} = this.props;
          const {mousePosition} = this.state;

          const targetAxis = pTargetAxis || axis;
          const axScroll = axisToProps[axis];
          const axTarget = axisToProps[targetAxis];

          const delta = [mousePosition[0] - coords[0], mousePosition[1] - coords[1]];

          const st: any = this.state;

          const {space: axisSpace, scrollSpace: axisScrollSpace}: { scrollSpace: number, space: number } = extractValues(st, axis);
          const {scrollSpace, space}: { scrollSpace: number, space: number } = extractValues(st, targetAxis);

          const scrollFactor =
            axis === targetAxis
              ? scrollSpace / space
              : (axisScrollSpace - axisSpace) / space;

          const barPosition: any = scrollableParent.getBoundingClientRect();
          if (this.state.barLocation === 'fixed') {
            const X = axis === 'vertical' ? st.scrollPosition[0] : st.scrollPosition[0] - delta[axTarget.coord] * scrollFactor;
            const Y = axis !== 'vertical' ? st.scrollPosition[1] : st.scrollPosition[1] - delta[axTarget.coord] * scrollFactor;
            window.scrollTo(X, Y);
          }
          else if (barPosition[axTarget.start] < coords[axTarget.coord] && barPosition[axTarget.end] > coords[axTarget.coord]) {
            scrollableParent[axScroll.scroll] = st.scrollPosition[axScroll.coord] - delta[axTarget.coord] * scrollFactor;
          }

        }
      })
      .start('init')
  }

  componentWillUnmount() {
    this.dragMachine.destroy();
    this.dettach();
  }

  componentDidUpdate(prevProps: IStrollerProps) {
    this.dragMachine.attrs({enabled: this.props.draggable});
    this.dragMachine.put('check');
    if (this.props.scrollKey !== prevProps.scrollKey) {
      this.onContainerScroll();
    }
  }

  private onContainerScroll = () => {
    const topNode = this.scrollableParent as any;

    const scrollTop = topNode.scrollTop;
    const scrollHeight = topNode.scrollHeight;
    const clientHeight = topNode.clientHeight;

    const scrollLeft = topNode.scrollLeft;
    const scrollWidth = topNode.scrollWidth;
    const clientWidth = topNode.clientWidth;

    const isFixed = this.state.barLocation === 'fixed';

    const st: any = this.state;

    const {axis = 'vertical', targetAxis} = this.props;

    const mainScroll = extractValues(st, axis);

    this.setState({
      scrollWidth,
      scrollHeight,

      clientWidth: isFixed ? window.innerWidth : clientWidth,
      clientHeight: isFixed ? window.innerHeight : clientHeight,

      scrollLeft: isFixed ? window.scrollX : scrollLeft,
      scrollTop: isFixed ? window.scrollY : scrollTop,

      hasScroll: mainScroll.scrollSpace > mainScroll.space,
    });
  };

  private attach(parent: HTMLElement | Window) {
    this.attachParent = parent;
    parent.addEventListener('scroll', this.onContainerScroll);
  }

  private dettach() {
    if (this.attachParent) {
      this.attachParent.addEventListener('scroll', this.onContainerScroll);
    }
  }

  private setScrollContainer = (ref: HTMLElement | null) => this.scrollContainer = ref;

  private setTopNode = (topNode: HTMLElement) => this.topNode = topNode;

  private setBarRef = (barRef: HTMLElement) => {
    this.barRef = barRef;
    this.dragMachine
      .attrs({node: this.barRef})
      .put('check');
  };

  private strollerProviderValue = {
    setScrollContainer: this.setScrollContainer
  };

  render() {
    const {
      children,
      bar,
      axis = 'vertical',
      targetAxis,
      oppositePosition = false,
      draggable = false,
      barSizeFunction,
      barClassName,
      SideBar,
    } = this.props;

    const {dragPhase} = this.state;
    const st: any = this.state;

    const ax = axisToProps[axis];

    const scrollSpace: number = st[ax.scrollSpace];

    const Bar = this.props.scrollBar || StollerBar;

    return (
      <StrollerStateProvider value={this.state}>
        <div ref={this.setTopNode as any} style={strollerStyle}>
          <StrollerProvider value={this.strollerProviderValue}>
            {children}
          </StrollerProvider>
          {scrollSpace && <Bar
            mainScroll={extractValues(st, axis)}
            targetScroll={extractValues(st, targetAxis || axis)}

            forwardRef={this.setBarRef}
            internal={bar}
            axis={axis}
            targetAxis={targetAxis || axis}

            oppositePosition={oppositePosition}

            draggable={draggable}
            dragging={dragPhase === 'dragging'}

            sizeFunction={barSizeFunction}
            location={st.barLocation}

            className={barClassName}
            SideBar={SideBar}
          />
          }
        </div>
      </StrollerStateProvider>
    );
  }
}