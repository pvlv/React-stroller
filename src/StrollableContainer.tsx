import {IContainerProps, Strollable, strollerStyle} from "./Container";
import {Stroller, IStrollerProps} from "./Stroller";
import * as React from "react";
import {StrollCaptor} from "./StrollCaptor";

export type Props = IContainerProps & IStrollerProps;

export const StrollableContainer: React.SFC<Props> = ({
                                                        children,
                                                        className,

                                                        axis,
                                                        bar,
                                                        scrollBar,
                                                        oppositePosition,
                                                        draggable,

                                                        barSizeFunction,


                                                        overrideLocation,
                                                        targetAxis,

                                                        overscroll,

                                                        scrollKey,
                                                        gap
                                                      }) => (
  <div style={strollerStyle}>
    <Stroller
      axis={axis}
      targetAxis={targetAxis}

      bar={bar}
      scrollBar={scrollBar}
      barSizeFunction={barSizeFunction}

      oppositePosition={oppositePosition}
      draggable={draggable}

      overrideLocation={overrideLocation}

      scrollKey={scrollKey}
    >
      <Strollable axis={axis} className={className} overscroll={overscroll} gap={gap}>
        <StrollCaptor/>
        {children}
      </Strollable>
    </Stroller>
  </div>
)