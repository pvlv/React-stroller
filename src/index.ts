import {Strollable} from "./Container";
import {Stroller} from './Stroller';
import {StrollableContainer} from "./StrollableContainer";
import {StrollCaptor} from './StrollCaptor';

export {
  Strollable,
  Stroller,
  StrollableContainer,
  StrollCaptor
}