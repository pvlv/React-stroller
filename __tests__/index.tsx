import * as React from 'react';
import { render } from '@testing-library/react';
import {Strollable, StrollableContainer} from '../src';

describe('Should render correctly', () => {
  it('Strollable', () => {
    const { container } = render(<Strollable/>);

    expect(container).toMatchSnapshot();
  });

  it('StrollableContainer', () => {
   const { container } = render(<StrollableContainer>child</StrollableContainer>);
      expect(container).toMatchSnapshot();
  });
});
