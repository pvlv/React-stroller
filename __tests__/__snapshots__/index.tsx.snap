// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly Strollable 1`] = `
<div>
  <div
    style="height: 100%; width: 100%; overflow: hidden; max-width: inherit; max-height: inherit;"
  >
    <div
      style="width: calc(100% + 0px); height: 100%; max-width: inherit; max-height: inherit; position: relative; overflow-y: scroll; padding-right: 24px; box-sizing: content-box;"
    >
      <div
        style="width: 100%; height: 100%; position: relative;"
      />
    </div>
  </div>
</div>
`;

exports[`Should render correctly StrollableContainer 1`] = `
<div>
  <div
    style="height: 100%; width: 100%; max-width: inherit; max-height: inherit;"
  >
    <div
      style="height: 100%; width: 100%; max-width: inherit; max-height: inherit;"
    >
      <div
        style="height: 100%; width: 100%; overflow: hidden; max-width: inherit; max-height: inherit;"
      >
        <div
          style="width: calc(100% + 0px); height: 100%; max-width: inherit; max-height: inherit; position: relative; overflow-y: scroll; padding-right: 24px; box-sizing: content-box;"
        >
          <div
            style="width: 100%; height: 100%; position: relative;"
          >
            <div />
            child
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</div>
`;
